pkg/cloud/vsphere/services/govmomi/service.go

```diff
@@ -135,19 +135,17 @@ func (vms *VMService) DestroyVM(ctx *context.MachineContext) (infrav1.VirtualMac
 		return vm, err
 	}
 
-	// check if VM actually exists
-	if ctx.VSphereMachine.Spec.MachineRef != "" {
-		moRefID, err := findVMByInstanceUUID(ctx)
-		if err != nil {
-			return vm, err
-		}
-		if moRefID == "" {
-			// No vm exists
-			// remove the MachineRef and set the vm state to notfound
-			ctx.VSphereMachine.Spec.MachineRef = ""
-			vm.State = infrav1.VirtualMachineStateNotFound
-			return vm, nil
-		}
+	// check if the vm existts
```

Comment:
Actually seems like there is an additional issue here
if the `ctx.VSphereMachine.Spec.MachineRef` is set then we should be getting the VM via the `getVMfromMachineRef` and not the `findVMByInstanceUUID(ctx)` directly. This is because in the case of pivot the `MachineRef` is the accurate reference, and in this situation, the `findVMByInstanceUUID` might not be the correct way to find the VM.

Reply:
Hi @sidharthsurana,

We're specifically avoiding the MoRef in this case to always do the instance UUID lookup. It's because this code removed the safeguards the old [`lookupVM` function](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/release-0.3/pkg/cloud/vsphere/services/govmomi/util.go#L171-L233) had to assert that a MoRef actually pointed to a real VM. The code @yastij wrote above bypasses the MoRef altogether to be certain of whether the VM still exists or not. This is an iterative step to removing the MoRef as the token in the `MachineRef` altogether. 

This was part of the service refactor, and I just didn't catch it when I reviewed that PR. We need to reintroduce the safeguards of `lookupVM`. If a MorRef was present, it was always checked to make sure it actually existed. It didn't just assume it did.

Reply:
Agreed! the previous logic in the lookupVM needs to be re-introduced!